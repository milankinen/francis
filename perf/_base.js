const memwatch = require("@airbnb/node-memwatch")
const Benchmark = require("benchmark")
const bytes = require("pretty-bytes")
const { format } = require("util")
const { Francis, Bacon, Kefir } = require("./_libs")

const MEMORY = !!process.env.SHOW_MEMORY_USAGE
let heapStats = []
let gcTime = 0

const withFrancis = f => (...args) => f(Francis, ...args)
const withBacon = f => (...args) => f(Bacon, ...args)
const withKefir = f => (...args) => f(Kefir, ...args)

const resetHeapStats = () => {
  gcTime = 0
  heapStats = []
}

const getAvgMemoryUsage = stats =>
  stats.length ? bytes(Math.floor(stats.reduce((a, s) => a + s, 0) / stats.length)) : "unknown"
const getMaxMemoryUsage = stats =>
  stats.length ? bytes(Math.floor(stats.reduce((a, s) => Math.max(a, s), 0))) : "unknown"

if (MEMORY) {
  memwatch.on("stats", stats => {
    heapStats.push(stats.used_heap_size)
    gcTime += stats.gc_time / 1000000.0
  })
}

const testCases = (label, fn, combinations = []) => {
  const accepted = (testFn, args) => !testFn.accept || testFn.accept(...args)
  const cases = (name, testFn) =>
    combinations
      .map(
        args =>
          accepted(testFn, args) && {
            description:
              `${name} ${format(label + "|||", ...args).split("|||")[0]}` +
              (testFn.note ? ` (${testFn.note})` : ""),
            defer: true,
            fn: d => {
              const done = () => d.resolve()
              testFn(...args, done)
            },
          },
      )
      .filter(x => x)
  return [
    ...(process.env.NO_FRANCIS ? [] : cases("francis", withFrancis(fn))),
    ...(process.env.NO_BACON ? [] : cases("bacon  ", withBacon(fn))),
    ...(process.env.NO_KEFIR || !fn.kefir ? [] : cases("kefir  ", withKefir(fn.kefir))),
  ]
}

const runTests = (suiteName, testCases) => {
  return new Promise(resolve => {
    console.log(`Running suite ${suiteName}...`)
    console.log("")
    const maxLen = testCases.reduce((m, t) => Math.max(t.description.length, m), 0)
    resetHeapStats()
    testCases
      .reduce(
        (suite, test) => suite.add(test.description.padEnd(maxLen + 1), test),
        new Benchmark.Suite(),
      )
      .on("complete", function() {
        if (MEMORY) {
          const stats = [...heapStats]
          console.log("Memory statistics:")
          console.log(" > Avg mem:", getAvgMemoryUsage(stats))
          console.log(" > Max mem:", getMaxMemoryUsage(stats))
          console.log(" > GC time:", gcTime.toFixed(2), "ms")
        } else {
          this.forEach(test => console.log(test.toString()))
        }
        console.log("")
        resolve()
      })
      .run({ async: true })
  })
}

module.exports = {
  testCases,
  runTests,
}
