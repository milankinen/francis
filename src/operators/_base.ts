import {
  NOOP_SUBSCRIBER,
  NOOP_SUBSCRIPTION,
  sendEndSafely,
  sendErrorSafely,
  sendEventSafely,
  sendInitialSafely,
  sendNoInitialSafely,
  Source,
  Subscriber,
  Subscription,
} from "../_core"
import { MAX_ORDER } from "../_priority"
import { Transaction } from "../_tx"
import { disableNoUnusedWarning } from "../_util"
import { Scheduler, Task } from "../scheduler/index"

export enum EventType {
  INITIAL = 1,
  NO_INITIAL,
  EVENT,
  ERROR,
  END,
}

/**
 * This function is meant for actual sources (e.g. once, fromArray etc..) to
 * enable multicasting (thus keeping the source's codebase clean)
 *
 * @param source Unicasted "bare" source
 */
export function identity<T>(source: Source<T>): Operator<T, T> {
  return new Identity(source)
}

export abstract class MulticastImplementation<T> implements Subscriber<T> {
  public readonly next: Subscriber<T> = NOOP_SUBSCRIBER

  /**
   * This hook is called every time when operator is activated (gets its first
   * subscriber). If this function returns false, operator aborts the current
   * subscription process. The impelementation may also queue tasks to the given
   * scheduler before returning.
   *
   * @param scheduler
   * @param subscriber
   */
  public beforeActivation(scheduler: Scheduler, subscriber: Subscriber<T>): boolean {
    return true
  }

  /**
   * This hook is called every time when operator is multicasted (gets a new subscriber
   * after initial activation). If this function returns false, operator aborts the current
   * subscription process. The impelementation may also queue tasks to the given
   * scheduler before returning.
   *
   * @param scheduler
   * @param subscriber
   */
  public beforeMulticast(scheduler: Scheduler, subscriber: Subscriber<T>): boolean {
    return true
  }

  public isActive(): boolean {
    return true
  }

  public initial(tx: Transaction, val: T): void {
    this.next.initial(tx, val)
  }

  public noinitial(tx: Transaction): void {
    this.next.noinitial(tx)
  }

  public event(tx: Transaction, val: T): void {
    this.next.event(tx, val)
  }

  public error(tx: Transaction, err: Error): void {
    this.next.error(tx, err)
  }

  public end(tx: Transaction): void {
    this.next.end(tx)
  }
}

const NOOP_MCAST = new class NoopImpl extends MulticastImplementation<any> {}()

export abstract class Operator<A, B>
  implements Subscriber<A>, Source<B>, AbortSubscriptionListener<B> {
  public readonly weight: number

  protected readonly order: number
  protected readonly source: Source<A>
  protected readonly next: MulticastImplementation<B> = NOOP_MCAST

  private __subs: Subscription = NOOP_SUBSCRIPTION
  private __active: boolean = false

  constructor(origin: Source<A>) {
    false && disableNoUnusedWarning(this.__handleDispose, this.__handeReorder)
    this.source = origin
    this.order = MAX_ORDER + 1
    this.weight = origin.weight
  }

  /**
   * Called only internally from Property and EventStream constructors
   * @param impl Multicast implementation to use for this operator
   */
  public setMulticastImplementation(impl: MulticastImplementation<B>): void {
    // tslint:disable-next-line:semicolon whitespace
    ;(this.next as MulticastImplementation<B>) = impl
  }

  /**
   * Called only internally when new EventStream or Property subscribes to this operator
   * during their activation.
   * @param scheduler
   * @param subscriber
   * @param order
   */
  public subscribe(scheduler: Scheduler, subscriber: Subscriber<B>, order: number): Subscription {
    const next = this.next.next
    return next === NOOP_SUBSCRIBER
      ? this.__handleActivation(scheduler, subscriber, order)
      : this.__handleMulticast(scheduler, subscriber, order)
  }

  public abstract initial(tx: Transaction, val: A): void

  public abstract event(tx: Transaction, val: A): void

  public noinitial(tx: Transaction): void {
    this.next.noinitial(tx)
  }

  public error(tx: Transaction, err: Error): void {
    this.next.error(tx, err)
  }

  public end(tx: Transaction): void {
    this.next.end(tx)
  }

  public isActive(): boolean {
    return this.__active
  }

  public handleAbort(subscriber: Subscriber<B>): void {}

  protected activateFirst(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    // TODO
    this.__active = true
    this.__subs = this.source.subscribe(scheduler, this, order)
    return new MulticastSubscription(subscriber, this as any)
  }

  protected activateLate(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    return new MulticastSubscription(subscriber, this as any)
  }

  protected abort(scheduler: Scheduler, subscriber: Subscriber<B>, order: number): Subscription {
    scheduler.scheduleAbortSubscription(new AbortSubscriptionTask(this, subscriber))
    return this.activateLate(scheduler, subscriber, order)
  }

  protected propagateReorder(order: number): void {
    this.__subs.reorder(order)
  }

  protected dispose(): void {
    this.__active = false
    this.__subs.dispose()
    this.__subs = NOOP_SUBSCRIPTION
  }

  protected handleActivation(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    return this.activateFirst(scheduler, subscriber, order)
  }

  protected handleMulticast(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    return this.activateLate(scheduler, subscriber, order)
  }

  protected handleReorder(order: number): void {
    this.propagateReorder(order)
  }

  protected handleDispose(): void {
    this.dispose()
  }

  private __updateNext(next: Subscriber<B>): void {
    // tslint:disable-next-line:semicolon whitespace
    ;(this.next.next as Subscriber<B>) = next
  }

  private __updateOrder(order: number): void {
    // tslint:disable-next-line:semicolon whitespace
    ;(this.order as number) = order
  }

  private __handleActivation(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    if (this.next.beforeActivation(scheduler, subscriber)) {
      this.__updateNext(subscriber)
      this.__updateOrder(order)
      return this.handleActivation(scheduler, subscriber, order)
    } else {
      return NOOP_SUBSCRIPTION
    }
  }

  private __handleMulticast(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    if (this.next.beforeMulticast(scheduler, subscriber)) {
      const next = this.next.next
      if (next instanceof MulticastDelegatee) {
        next.add(subscriber, order)
      } else {
        this.__updateNext(
          new MulticastDelegatee({ s: next, o: this.order }, { s: subscriber, o: order }),
        )
      }
      if (order < this.order) {
        this.__updateOrder(order)
        this.handleReorder(order)
      }
      return this.handleMulticast(scheduler, subscriber, order)
    } else {
      return NOOP_SUBSCRIPTION
    }
  }

  private __handleDispose(subscriber: Subscriber<B>): void {
    const next = this.next.next
    if (next instanceof MulticastDelegatee) {
      const { o: order, s: updated } = next.remove(subscriber)
      this.__updateNext(updated)
      if (order !== this.order) {
        this.__updateOrder(order)
        this.handleReorder(order)
        this.__subs.reorder(order)
      }
    } else {
      this.__updateNext(NOOP_SUBSCRIBER)
      this.__updateOrder(MAX_ORDER + 1)
      this.handleDispose()
    }
  }
  private __handeReorder(subscriber: Subscriber<B>, order: number): void {
    const next = this.next.next
    next instanceof MulticastDelegatee && next.updateOrder(subscriber, order)
    if (order < this.order) {
      this.__updateOrder(order)
      this.handleReorder(order)
    }
  }
}

export interface AbortSubscriptionListener<T> {
  handleAbort(subsciber: Subscriber<T>): void
}

export class AbortSubscriptionTask<T> implements Task {
  constructor(private li: AbortSubscriptionListener<T>, private subscriber: Subscriber<T>) {}

  public run(): void {
    this.li.handleAbort(this.subscriber)
  }
}

const NIL: OrderedSubscriberNode<any> = {
  a: false,
  n: null as any,
  o: MAX_ORDER + 1,
  s: NOOP_SUBSCRIBER,
}

class MulticastDelegatee<T> implements Subscriber<T> {
  private n: number
  private head: OrderedSubscriberNode<T>
  private tail: OrderedSubscriberNode<T>

  constructor(a: OrderedSubscriber<T>, b: OrderedSubscriber<T>) {
    this.head = this.tail = NIL
    this.head = { ...a, a: true, n: this.head }
    this.head = { ...b, a: true, n: this.head }
    this.n = 2
  }

  public add(subscriber: Subscriber<T>, order: number): void {
    this.head = { o: order, s: subscriber, a: true, n: this.head }
    ++this.n
  }

  public remove(subscriber: Subscriber<T>): OrderedSubscriber<T> {
    let prev = NIL
    let next = this.head
    let order = MAX_ORDER + 1
    while (next !== NIL) {
      if (next.s === subscriber) {
        next.a = false
        prev !== NIL ? (prev.n = next.n) : (this.head = next.n)
        --this.n
      } else if (next.o < order) {
        order = next.o
      }
      prev = next
      next = next.n
    }
    return {
      o: order,
      s: this.n > 1 ? this : this.head.s,
    }
  }

  public updateOrder(subscriber: Subscriber<T>, order: number): void {
    let next = this.head
    while (next !== NIL) {
      if (next.s === subscriber) {
        next.o = order
        return
      } else {
        next = next.n
      }
    }
  }

  public isActive(): boolean {
    return true
  }

  public initial(tx: Transaction, val: T): void {
    let next = this.head
    while (next !== NIL) {
      next.a && sendInitialSafely(tx, next.s, val)
      next = next.n
    }
  }

  public noinitial(tx: Transaction): void {
    let next = this.head
    while (next !== NIL) {
      next.a && sendNoInitialSafely(tx, next.s)
      next = next.n
    }
  }

  public event(tx: Transaction, val: T): void {
    let next = this.head
    while (next !== NIL) {
      next.a && sendEventSafely(tx, next.s, val)
      next = next.n
    }
  }

  public error(tx: Transaction, err: Error): void {
    let next = this.head
    while (next !== NIL) {
      next.a && sendErrorSafely(tx, next.s, err)
      next = next.n
    }
  }

  public end(tx: Transaction): void {
    let next = this.head
    while (next !== NIL) {
      next.a && sendEndSafely(tx, next.s)
      next = next.n
    }
  }
}

interface OrderedSubscriberNode<T> extends OrderedSubscriber<T> {
  n: OrderedSubscriberNode<T>
  a: boolean
}

class MulticastSubscription<T> implements Subscription {
  private s: Subscriber<T>
  private o: MulticastSubscriptionOwner<T>

  // tslint:disable-next-line:ban-types
  constructor(subscriber: Subscriber<T>, owner: MulticastSubscriptionOwner<T>) {
    this.s = subscriber
    this.o = owner
  }

  public dispose(): void {
    this.o.__handleDispose(this.s)
  }

  public reorder(order: number): void {
    this.o.__handeReorder(this.s, order)
  }
}

interface MulticastSubscriptionOwner<T> {
  __handleDispose(subscriber: Subscriber<T>): void
  __handeReorder(subscriber: Subscriber<T>, order: number): void
}

interface OrderedSubscriber<T> {
  s: Subscriber<T>
  o: number
}

class Identity<T> extends Operator<T, T> {
  public initial(tx: Transaction, val: T): void {
    this.next.initial(tx, val)
  }
  public event(tx: Transaction, val: T): void {
    this.next.event(tx, val)
  }
}
