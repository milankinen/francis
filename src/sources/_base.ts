import { __DEVELOPER__, logAndThrow } from "../_assert"
import {
  NOOP_SUBSCRIBER,
  sendRootEnd,
  sendRootError,
  sendRootNext,
  Source,
  Subscriber,
  Subscription,
} from "../_core"
import { AnyEvent } from "../_interfaces"
import { isEnd, isError, isEvent, isNext } from "../Event"
import { scheduleActivationTask, Task } from "../scheduler/index"

export abstract class Root<T> implements Source<T> {
  public readonly weight: number = 1
  public ended: boolean = false
  constructor(public readonly sync: boolean) {}

  public subscribe(subscriber: Subscriber<T>, order: number): Subscription {
    return this.create(subscriber)
  }

  protected abstract create(subscriber: Subscriber<T>): Activation<T, any>
}

export abstract class Activation<T, R extends Root<T>> implements Task, Subscription {
  // the real subscriber
  protected readonly real: Subscriber<T>
  protected readonly owner: R
  protected active: boolean = true
  protected subscriber: Subscriber<T>

  constructor(owner: R, subscriber: Subscriber<T>) {
    this.real = this.subscriber = subscriber
    this.owner = owner
  }

  public run(): void {
    if (this.active) {
      if (this.owner.ended) {
        sendRootEnd(this.subscriber)
      } else {
        if (this.subscriber.begin()) {
          this.start()
        }
      }
    }
  }

  public activate(): void {
    if (this.owner.sync) {
      this.run()
    } else {
      scheduleActivationTask(this)
    }
  }

  public dispose(): void {
    this.active = false
    this.subscriber = NOOP_SUBSCRIBER
    this.stop()
  }

  public reorder(order: number): void {
    // no-op
  }

  protected abstract start(): void

  protected abstract stop(): void

  protected send(event: T | AnyEvent<T>): void {
    if (isEvent(event)) {
      if (isNext(event)) {
        this.sendNext(event.value)
      } else if (isError(event)) {
        this.sendError(event.error)
      } else if (isEnd(event)) {
        this.sendEnd()
      } else {
        if (__DEVELOPER__) {
          logAndThrow("**BUG** event type not known: " + event)
        }
      }
    } else {
      sendRootNext(this.subscriber, event)
    }
  }

  protected sendNext(val: T): void {
    sendRootNext(this.subscriber, val)
  }

  protected sendError(err: Error): void {
    sendRootError(this.subscriber, err)
  }

  protected sendEnd(): void {
    this.owner.ended = true
    sendRootEnd(this.subscriber)
  }
}
