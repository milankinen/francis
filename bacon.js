export * from "./dist/francis.bacon"
