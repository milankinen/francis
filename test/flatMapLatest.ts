import * as F from "../bacon"
import { runner, Sync } from "./_base"

describe("EventStream.flatMapLatest", () => {
  it("results in EventStream", () => {
    expect(F.once("tsers").flatMapLatest(x => x + "!")).toBeInstanceOf(F.EventStream)
  })

  it("uses the events from the latest inner stream only", done => {
    runner()
      .setup(record => {
        const a = F.sequentially(5, ["1", "2", "3"])
        const b = F.sequentially(5, ["a", "b", "c"])
        F.sequentially(12, [1, 2])
          .flatMapLatest(i => (i === 1 ? a : b))
          .subscribe(record)
      })
      .after(rec => expect(rec).toMatchSnapshot())
      .run(done)
  })

  it("does not re-activate inner stream if it does not change", done => {
    runner()
      .setup((record, wait) => {
        const inner = F.fromBinder(_ => {
          record("inner activated")
          return () => record("inner disposed")
        }).toProperty("tsers")
        const dispose = F.sequentially(1, [1, 2, 3, 4])
          .flatMapLatest(_ => inner)
          .subscribe(() => {})
        wait(50, dispose)
      })
      .after(rec => expect(rec).toEqual(["inner activated", "inner disposed"]))
      .run(done)
  })

  it("does not end if the latest inner stream does not end", done => {
    runner()
      .setup(record => {
        F.once("tsers")
          .flatMapLatest(msg => neverEnding(msg))
          .subscribe(record)
      })
      .after(rec => expect(rec).toMatchSnapshot())
      .run(done)
  })

  it("does not end if the outer stream does not end", done => {
    runner()
      .setup(record => {
        neverEnding("tsers")
          .flatMapLatest(msg => F.once(msg))
          .subscribe(record)
      })
      .after(rec => expect(rec).toMatchSnapshot())
      .run(done)
  })

  it("disposes previous inner stream properly", done => {
    runner()
      .setup(record => {
        const makeInner = (num: number) =>
          F.fromBinder(_ => {
            record("activated: " + num)
            return () => record("disposed: " + num)
          })
        F.sequentially(1, [1, 2, 3, 4])
          .flatMapLatest(i => makeInner(i))
          .subscribe(() => {})
      })
      .after(rec =>
        expect(rec).toEqual([
          "activated: 1",
          "disposed: 1",
          "activated: 2",
          "disposed: 2",
          "activated: 3",
          "disposed: 3",
          "activated: 4",
          // 4th inner stream is not disposed because outer stream ends
        ]),
      )
      .run(done)
  })
})

describe("Property.flatMapLatest", () => {
  it("results in Property", () => {
    expect(F.constant("tsers").flatMapLatest(x => x + "!")).toBeInstanceOf(F.Property)
  })

  it("emits its inner events synchronously", done => {
    runner()
      .setup(record => {
        F.constant("tsers")
          .flatMapLatest(msg => F.constant(msg + "!"))
          .subscribe(record)
        record(Sync)
      })
      .after(rec => expect(rec).toMatchSnapshot())
      .run(done)
  })

  it("emits initial value only once", done => {
    runner()
      .setup(record =>
        F.once(1)
          .toProperty(0)
          .flatMapLatest(x => x)
          .subscribe(record),
      )
      .after(rec => expect(rec).toMatchSnapshot())
      .run(done)
  })

  it("does not emit initial value if inner observable does not emit initial valu", done => {
    runner()
      .setup(record =>
        F.constant("tsers")
          .flatMapLatest(msg => F.once(msg).toProperty())
          .subscribe(record),
      )
      .after(rec => expect(rec).toMatchSnapshot())
      .run(done)
  })

  it("works with same origin (follows transaction semantics)", done => {
    runner()
      .setup(record => {
        const prop = F.sequentially(1, ["a", "b", "c"]).toProperty(">")
        prop.flatMapLatest(_ => prop).subscribe(record)
      })
      .after(rec => expect(rec).toMatchSnapshot())
      .run(done)
  })

  it("works with derived same origin (follows transaction semantics)", done => {
    runner()
      .setup(record => {
        const prop = F.sequentially(1, ["a", "b", "c"]).toProperty()
        prop.flatMapLatest(letter => prop.map(l => l.toUpperCase() + letter)).subscribe(record)
      })
      .after(rec => expect(rec).toMatchSnapshot())
      .run(done)
  })

  it("works with nested combines (follows transaction semantics)", done => {
    runner()
      .setup(record => {
        const lower = F.sequentially(2, ["a", "b"]).toProperty(">")
        const upper = lower.map(l => l.toUpperCase())
        const under = lower.map(l => "_" + l + "_")
        const all = lower.flatMapLatest(l => F.combineAsArray(l, upper, under))
        F.combineAsArray(upper, all).subscribe(record)
      })
      .after(rec => expect(rec).toMatchSnapshot())
      .run(done)
  })
})

function neverEnding<T>(val: T): F.EventStream<T> {
  return F.fromBinder(sink => {
    sink(val)
  })
}
