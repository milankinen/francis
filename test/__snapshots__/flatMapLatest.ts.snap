// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventStream.flatMapLatest does not end if the latest inner stream does not end 1`] = `
Array [
  Next {
    "value": "tsers",
  },
]
`;

exports[`EventStream.flatMapLatest does not end if the outer stream does not end 1`] = `
Array [
  Next {
    "value": "tsers",
  },
]
`;

exports[`EventStream.flatMapLatest uses the events from the latest inner stream only 1`] = `
Array [
  Next {
    "value": "1",
  },
  Next {
    "value": "2",
  },
  Next {
    "value": "a",
  },
  Next {
    "value": "b",
  },
  Next {
    "value": "c",
  },
  End {},
]
`;

exports[`Property.flatMapLatest does not emit initial value if inner observable does not emit initial valu 1`] = `
Array [
  Next {
    "value": "tsers",
  },
  End {},
]
`;

exports[`Property.flatMapLatest emits initial value only once 1`] = `
Array [
  Next {
    "value": 0,
  },
  Next {
    "value": 1,
  },
  End {},
]
`;

exports[`Property.flatMapLatest emits its inner events synchronously 1`] = `
Array [
  Next {
    "value": "tsers!",
  },
  End {},
  Sync {},
]
`;

exports[`Property.flatMapLatest works with derived same origin (follows transaction semantics) 1`] = `
Array [
  Next {
    "value": "Aa",
  },
  Next {
    "value": "Bb",
  },
  Next {
    "value": "Cc",
  },
  End {},
]
`;

exports[`Property.flatMapLatest works with nested combines (follows transaction semantics) 1`] = `
Array [
  Next {
    "value": Array [
      ">",
      Array [
        ">",
        ">",
        "_>_",
      ],
    ],
  },
  Next {
    "value": Array [
      "A",
      Array [
        "a",
        "A",
        "_a_",
      ],
    ],
  },
  Next {
    "value": Array [
      "B",
      Array [
        "b",
        "B",
        "_b_",
      ],
    ],
  },
  End {},
]
`;

exports[`Property.flatMapLatest works with same origin (follows transaction semantics) 1`] = `
Array [
  Next {
    "value": ">",
  },
  Next {
    "value": "a",
  },
  Next {
    "value": "b",
  },
  Next {
    "value": "c",
  },
  End {},
]
`;
